### PR TITLE
Add support for LLVM 16

### PR DESCRIPTION
**Pull request by linehill**
* Switch to use opaque pointers for LLVM 16+.
* Adjust LLVM passes for LLVM 16.
* Add a workaround for unsupported __bf16 type.
* Update README.md.

TODOs before ready for review:

- [x] Wait until LLVM 16 release branch stabilizes.
- [x] Wait until llvm-spirv branches for LLVM 16.
- [x] Install LLVM 16 on CI
- [x] Merge #385 
- [x] Run unit tests again
- [x] If not resolved, exclude failing tests

### CODE REVIEW

**Comment by pvelesko** [comment]
>  Should we now add an LLVM 16 test env to the CI?

I'll compile & install clang 16. Any specific test in mind or should we just switch the compiler?

**Comment by linehill** [comment]
We probably want to support `N` latest LLVM versions so it would be preferable if we test the CHIP-SPV against all the supported LLVM versions.

**Comment by pvelesko** [comment]
so run all current unit tests for N releases of clang? that's a lot of unit tests. Maybe I can add a workflow to test older versions right before merging into main.

**Comment by pjaaskel** [comment]
I'd support only the last two releases, but Argonne might want to support some older versions. So my proposal is to test the latest LLVM version and what Argonne needs more thoroughly, and the rest with more of a "smoke test"-attitude.

**Comment by pvelesko** [comment]
```
Consolidate compiler generated dependencies of target jnf
[ 34%] Building CXX object catch/catch_tests/unit/deviceLib/SinglePrecisionMathFunctions/CMakeFiles/jnf.dir/jnf.cc.o
InvalidArraySize: Array size must be at least 1: [0 x float]
clang-16: error: hipspv-link command failed with exit code 9 (use -v to see invocation)

failed to execute:/home/pvelesko/install/clang/16.0/bin/clang++ -D__HIP_PLATFORM_SPIRV__= -D__HIP_PLATFORM_SPIRV__= --offload=spirv64 -nohipwrapperinc --hip-path=/home/pvelesko/CHIP-SPV/chip-spv/build --target=x86_64-unknown-linux-gnu -include /home/pvelesko/CHIP-SPV/chip-spv/build/include/hip/spirv_fixups.h -isystem "/home/pvelesko/CHIP-SPV/chip-spv/build/include" -x hip --target=x86_64-linux-gnu -I//home/pvelesko/CHIP-SPV/chip-spv/build/include -DKERNELS_PATH="\"/home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/kernels/"\" -I/home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/external/Catch2 -I/home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/./include -I/home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/./kernels -I/home/pvelesko/CHIP-SPV/chip-spv/build/include -I/home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/external/picojson -I/include -Wno-duplicate-decl-specifier -Wno-tautological-constant-compare -Wno-c++20-extensions -Wno-unused-result -Wno-delete-abstract-non-virtual-dtor -Wno-deprecated-declarations -Wunused-command-line-argument -gdwarf-4 --std=c++17 -fPIE -Wno-format-extra-args -mf16c -std=c++17 -MD -MT catch/catch_tests/unit/deviceLib/SinglePrecisionMathFunctions/CMakeFiles/jnf.dir/jnf.cc.o -MF CMakeFiles/jnf.dir/jnf.cc.o.d -o CMakeFiles/jnf.dir/jnf.cc.o -c /home/pvelesko/CHIP-SPV/chip-spv/HIP/tests/catch/unit/deviceLib/SinglePrecisionMathFunctions/jnf.cc
make[3]: *** [catch/catch_tests/unit/deviceLib/SinglePrecisionMathFunctions/CMakeFiles/jnf.dir/build.make:76: catch/catch_tests/unit/deviceLib/SinglePrecisionMathFunctions/CMakeFiles/jnf.dir/jnf.cc.o] Error 9
make[2]: *** [CMakeFiles/Makefile2:9167: catch/catch_tests/unit/deviceLib/SinglePrecisionMathFunctions/CMakeFiles/jnf.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:6610: catch/CMakeFiles/build_tests.dir/rule] Error 2
make: *** [Makefile:2162: build_tests] Error 2
```

**Comment by linehill** [comment]
Looking into this.

**Comment by linehill** [comment]
Fixed for LLVM 16. Need to check LLVM 15 as the `InvalidArraySize` error may appear on the earlier LLVM versions too.

File: cmake/UnitTests.cmake [hunk 1]
```diff
@@ -391,6 +391,13 @@ list(APPEND CPU_OPENCL_FAILED_TESTS "Unit_hipMallocPitch_KernelLaunch - float")
 list(APPEND CPU_OPENCL_FAILED_TESTS "Unit_hipMallocPitch_KernelLaunch - double") # Failed
 list(APPEND CPU_OPENCL_FAILED_TESTS "Unit_hipGraphAddHostNode_ClonedGraphwithHostNode") # Failed
 list(APPEND CPU_OPENCL_FAILED_TESTS "syncthreadsExitedThreads") # Timeout
+# The following warning appears in the CI with LLVM 16:
+#
+#   warning: stack frame size (18446744073709551608) exceeds limit (4294967295) in function '_Z15matrixTransposePfS_i'
+#
+# altought the kernels in the SPIR-V code do not have stack objects at
+# all and, otherwise, there does not seem to be anything out of ordinary.
+list(APPEND CPU_OPENCL_FAILED_TESTS "dynamic_shared" "hipDynamicShared" "cuda-template") # SEGFAULT
```
Comment (pvelesko) [inline cmake/UnitTests.cmake:400]: Where is this segfault coming from if "there does not seem to be anything out of ordinary."?

**Review by pvelesko — APPROVED** [review]
Just need clarification on test exclusion.

File: cmake/UnitTests.cmake (hunk 1 above)
Comment (linehill) [reply cmake/UnitTests.cmake:400]: No idea. All I know is that the segfault appears in the CI along with the warning message.

Comment (pvelesko) [reply cmake/UnitTests.cmake:400]: I'm not sure if just adding more and more issues into the excluded test list is a good idea. Main already seems to be broken I think we should resolve issues in main or at the very least not introduce any more bugs.

**Review by pvelesko — APPROVED** [review]
